src/main/java/bio/terra/dao/DatasetTableDao.java

```diff
@@ -1,17 +1,90 @@
+import java.util.stream.Collectors;
+
 @Repository
 public class DatasetTableDao extends TableDaoBase {
 
+    private static final Logger logger = LoggerFactory.getLogger(DatasetTableDao.class);
+
+    private static final String sqlInsertTable = "INSERT INTO dataset_table " +
+        "(name, dataset_id, primary_key) VALUES (:name, :dataset_id, :primary_key)";
+
+    private final DataSource jdbcDataSource;
     private final NamedParameterJdbcTemplate jdbcTemplate;
 
+
     @Autowired
-    public DatasetTableDao(NamedParameterJdbcTemplate jdbcTemplate) {
-        super(jdbcTemplate, "dataset_table", "dataset_column", "dataset_id");
-        this.jdbcTemplate = jdbcTemplate;
+    public DatasetTableDao(DataRepoJdbcConfiguration dataRepoJdbcConfiguration) {
+        super(dataRepoJdbcConfiguration, "dataset_table", "dataset_column", "dataset_id");
+        this.jdbcDataSource = dataRepoJdbcConfiguration.getDataSource();
+        this.jdbcTemplate = new NamedParameterJdbcTemplate(this.jdbcDataSource);
+    }
+
+    @Override
+    protected UUID createTable(UUID parentId, Table table) {
+        MapSqlParameterSource params = new MapSqlParameterSource();
+        DaoKeyHolder keyHolder = new DaoKeyHolder();
+
+        logger.info(parentId.toString());
+        params.addValue("name", table.getName());
+        params.addValue("dataset_id", parentId);
+
+        List<String> naturalKeyStringList = table.getPrimaryKey()
+            .orElseThrow(() -> new IllegalStateException("Dataset table " + table.getName() + " has no primary key"))
+            .stream()
+            .map(Column::getName)
+            .collect(Collectors.toList());
+        try (Connection connection = jdbcDataSource.getConnection()) {
+            params.addValue("primary_key", DaoUtils.createSqlStringArray(connection, naturalKeyStringList));
+        } catch (SQLException e) {
+            logger.error("Failed to convert primary key list to SQL array", e);
+            throw new IllegalArgumentException("Failed to convert primary key list to SQL array", e);
+        }
+
+        jdbcTemplate.update(sqlInsertTable, params, keyHolder);
+        return keyHolder.getId();
+    }
+
+    @Override
+    protected Table retrieveTable(ResultSet rs) throws SQLException {
+        Table table = new Table()
+            .id(rs.getObject("id", UUID.class))
+            .name(rs.getString("name"));
+
+        List<String> primaryKey = DaoUtils.getStringList(rs, "primary_key");
+        List<Column> columns = retrieveColumns(table);
+        Map<String, Column> columnMap = columns
+            .stream()
+            .collect(Collectors.toMap(Column::getName, Function.identity()));
+
+        List<Column> naturalKeyColumns = Optional.ofNullable(primaryKey)
+            .orElse(Collections.emptyList())
+            .stream()
+            .map(columnMap::get)
+            .collect(Collectors.toList());
+        table.primaryKey(naturalKeyColumns);
+
+        return table.columns(columns);
```

Comment:
This is a lot of code to add in DatasetTableDao and SnapshotTableDao. If I had done it, I would have put the primary key into the TableBaseDao and let it be optional.

If the consensus is, that this layering is the way to go, then we should think about whether it would be better to stop sharing any of the base class. Using overridden protected methods maybe makes things more obscure than a few tidbits of duplicate code.